Update USACO dec18-bucketlist.mdx (#1511)
cpinitiative/usaco-guide: solutions/USACO dec18-bucketlist.mdx

id: usaco-856
source: USACO Bronze 2018 December
title: The Bucket List

> author: Ananth Kashyap (Python) & Sathvik Chundru (Cpp)

Reviewer: ```suggestion
author: Ananth Kashyap (Python) and Sathvik Chundru (C++)
```

</PySection>

> </LanguageSection>
> 
> <LanguageSection>

Reviewer 2: ```suggestion
```

Reviewer 3: the two solutions are different, so i think it's better to keep them separated?

Author: thats what i was thinking

Author: also it passes all the checks this way

<CPPSection>

```cpp
#include <iostream>

> #include <algorithm>

Reviewer 3: don't you need cstdio for freopen?

Reviewer 2: iostream has freopen too

Reviewer 2: iostream also has min/max too

Reviewer 2: ok so ignore my comments above; add `#include <cstdio>`

Reviewer 3: address this

Author: done i removed algorithm

using namespace std;

> int event [1001];

Reviewer 3: ```suggestion
int event[1001];
```

</PySection>

</LanguageSection>

<LanguageSection>

<CPPSection>

```cpp
#include <iostream>
#include <algorithm>
using namespace std;
int event[1001]; // declaring the array globaly to intialize all the elements to 0
int main() {
	freopen("blist.in", "r", stdin);
	freopen("blist.out", "w", stdout); 
	int maxtime = 1000;
	int n; cin >> n;
		
		
	for (int i = 0; i < n; i++) { // iterating through n lines to input the data in
		int start; cin >> start; 
		int end; cin >> end;
		int amt; cin >> amt;
		event[start] += amt; 
		// simulating the events by adding/subtracting the amount from si and ti

> 		event[end] -= amt; // ^

Reviewer 3: what are si and ti?

Author: from the problem its what they call the input

Reviewer 3: aren't si and ti the `start` and `end` in your code? you're not adding/subtracting from `start` and `end`...

	}
		
		
	int res = 0, cur = 0; // keeping track of buckets being used starting from 0
	for (int i = 0; i <= maxtime; i++) {
		cur += event[i]; // adding the current bucket to event[i]
		res = max(res, cur); 

> 		// saving the curr in increasing order to get the 

Reviewer 3: ```suggestion
		// saving the cur in increasing order to get the 
```

</PySection>

</LanguageSection>

<LanguageSection>

<CPPSection>

```cpp
#include <iostream>
#include <algorithm>  
#include <cstdio>
using namespace std;

> int event[1001]; // declaring the array globaly to intialize all the elements to 0

Reviewer: ```suggestion
int event[1001]; // declaring the array globally to initialize all the elements to 0
```

Reviewer: Also, can we have a better name than `event` :P (or alternatively explain what event is)

int main() {
	freopen("blist.in", "r", stdin);
	freopen("blist.out", "w", stdout); 
	int maxtime = 1000;
	int n; cin >> n;

> 		
> 		

Reviewer: ```suggestion
		
```

> 	for (int i = 0; i < n; i++) { // iterating through n lines to input the data in

Reviewer: ```suggestion
	for (int i = 0; i < n; i++) { // iterating through every cow
```

		int start; cin >> start; 
		int end; cin >> end;
		int amt; cin >> amt;
		event[start] += amt; 
		// simulating the events by adding/subtracting the amount from si and ti

> 		event[end] -= amt; // ^

Reviewer: I feel like this should still be explained a bit more (either through more comments or through a brief editorial or both). I don't think Bronze people are familiar with this "prefix sum" ish sweep technique. What an "event" is isn't explained anywhere, for example

	}
		
		

> 	int res = 0, cur = 0; // keeping track of buckets being used starting from 0

Reviewer: I think the comment is a bit unclear -- what exactly does res and cur store?

	for (int i = 0; i <= maxtime; i++) {

> 		cur += event[i]; // adding the current bucket to event[i]

Reviewer: I think the comment can be improved -- comments literally explaining what the code does (for example, "adds event[i] to cur") are not very helpful. Ideally the comment should explain why we are doing this step (ie. "At time i, the number of buckets in use changes by `event[i]` so we want to update the number of buckets we are currently using accordingly" or something like that)

		res = max(res, cur); 
		// saving the cur in increasing order to get the 

> 		// highest value we can for the current bucket

Reviewer: This might be personal preference but I think it's more clear if the comments go before the code rather than after the code? (for consistency can we have comments appear before the code rather than after the code?)

</PySection>

</LanguageSection>

## Solution 2

Similar to the second solution in the official editorial. 

<LanguageSection>

<CPPSection>

```cpp
#include <iostream>
#include <algorithm>  
#include <cstdio>
using namespace std;
int event[1001]; // declaring the array globally to initialize all the elements to 0
int main() {
	freopen("blist.in", "r", stdin);
	freopen("blist.out", "w", stdout); 
	int maxtime = 1000;
	int n; cin >> n;
		
		
	for (int i = 0; i < n; i++) { // iterating through every cow
		int start; cin >> start; 
		int end; cin >> end;
		int amt; cin >> amt;
		// at time start, cow i needs to be milked, so 
		// we'll need amt additional buckets (hence adding amt).
		// at time end, cow i is done milking, so we can free up 

> 		// amt buckets (hence subtracting amt)

Reviewer 3: maybe you could put the first two lines above `event[start] += amt;` and the last two lines above `event[end] -= amt;` ?

Author: i think that would mess up the flow

Author: gonna do it then if lunchbox agrees as well

Reviewer 2: > maybe you could put the first two lines above `event[start] += amt;` and the last two lines above `event[end] -= amt;` ?

i agree

</PySection>

</LanguageSection>

> # Solution 2

Reviewer: ```suggestion
## Solution 2
```

(for consistency with the other header)

Similar to the second solution in the official editorial. 

<LanguageSection>

<CPPSection>

```cpp
#include <iostream>
#include <algorithm>  
#include <cstdio>
using namespace std;
int event[1001]; // declaring the array globally to initialize all the elements to 0
int main() {
	freopen("blist.in", "r", stdin);
	freopen("blist.out", "w", stdout); 
	int maxtime = 1000;
	int n; cin >> n;
		
		
	for (int i = 0; i < n; i++) { // iterating through every cow
		int start; cin >> start; 
		int end; cin >> end;
		int amt; cin >> amt;
		// at time start, cow i needs to be milked, so 
		// we'll need amt additional buckets (hence adding amt). 
		// at time end, cow i is done milking, so we can free up 

> 		// amt buckets (hence deleted amt). This technique is called a 

Reviewer: deleted ==> subtracting?

		// "sweep", because if you think of all the points
		// on a horizontal number line, we are "sweeping" across 

> 		// the line, processing each point that we come across.

Reviewer: Is it worth keeping the "sweep" stuff in, or should we remove it? It's already described in the official editorial and it isn't necessary to understand the code, so I feel like it's sort of out of place here / it might confuse people.

@maggie-j-liu @dongliu0426

Reviewer 2: yea i think "sweep" is kinda confusing for bronze

Reviewer 3: ^ i agree

if there's this many comments, they should probably be made into an editorial since it clutters the code

Author: gonna fix it

Reviewer 2: > if there's this many comments, they should probably be made into an editorial since it clutters the code

^

		event[start] += amt;
		event[end] -= amt; 
	}
		
	// res represents the maximum number of buckets we'll need 

> 	// to use at any given point in time and cur represents 

Reviewer: ```suggestion
	// to use at any given time and cur represents 
```
